### PR TITLE
Skip `scheduler.step()` if `optimizer.step()` isn't called in the iteration

### LINKED ISSUE #5558: Mixed precision: scheduler and optimizer are called in the wrong order
## 🐛 Bug

When using mixed-precision training, scheduler and optimizer are called in the wrong order. Warning is generated:

```
UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.
```

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1G7pk6E9XUYq-pS41DXKhqM9Srx8sikiP?usp=sharing

There are four tests. Three of them doesn't raise the warning:

1. test_amp_scheduler(precision=16, configure_optimizers=configure_optimizers_1)
2. test_amp_scheduler(precision=32, configure_optimizers=configure_optimizers_1)
3. test_amp_scheduler(precision=32, configure_optimizers=configure_optimizers_2)

This testcase raises the warning:

1. test_amp_scheduler(precision=16, configure_optimizers=configure_optimizers_2)

### To Reproduce

1. Create model with `configure_optimizers` in a following dictionary style:

```python
def configure_optimizers_2(model):
    optimizer = torch.optim.SGD(model.layer.parameters(), lr=0.1)
    scheduler = {'scheduler':  torch.optim.lr_scheduler.StepLR(optimizer, step_size=1),
              'name': 'learning_rate',
              'interval':'step',
              'frequency': 1}
    
    return {"optimizer": optimizer, "lr_scheduler": scheduler}
```

2. Enable mixed-precision training by setting `precision=16` in a `Trainer`
3. Start training

#### Note

When scheduler is defined in another way, the issue seems to not occur:

```python
def configure_optimizers_1(model):
    optimizer = torch.optim.SGD(model.layer.parameters(), lr=0.1)
    scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
    
    return {"optimizer": optimizer, "lr_scheduler": scheduler}
```

### Expected behavior

No warning

### Environment

* CUDA:
	- GPU:
		- Tesla P100-PCIE-16GB
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.1.4
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Fixes #5558

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @justusschock @awaelchli @akihironitta

### CODE REVIEW

**Comment by aleSuglia** [comment]
@akihironitta A fix is here ready: https://github.com/aleSuglia/pytorch-lightning/tree/bugfix/scheduler-before-optimizer

Google colab to test it: https://colab.research.google.com/drive/1nQOJ9K3JE0MtrtT5qWOxQymJfa74qGmK?usp=sharing

**Review by akihironitta — COMMENTED** [review]
@tchaton Not sure if my comment below makes sense (or if the fix is even in the right direction), but I'd love hear how I can proceed forward in general.

Feel free to push commits if you need.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 1]
```diff
@@ -476,6 +476,10 @@ def _update_learning_rates(
             if update_plateau_schedulers ^ lr_scheduler["reduce_on_plateau"]:
                 continue
 
+            # skip if `optimizer.step()` is skipped
+            if getattr(lr_scheduler["scheduler"].optimizer, "skipped_optimizer_step", False):
+                continue
+
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:482]: Yes, `skipped_optimizer_step` seems quite a hacky solution. Might be simpler to keep track of this within an other component like the Trainer

**Comment by akihironitta** [comment]
still wip
